File: DSA#/Sorting-Algos/CountingSort.cpp
#include <iostream>
#include <algorithm>
using namespace std;

void countingSort(int arr[], int n) {
    int max = *max_element(arr, arr + n);
    int count[max + 1] = {0};

    for (int i = 0; i < n; i++)
        count[arr[i]]++;

    int index = 0;
    for (int i = 0; i <= max; i++)
        while (count[i]--)
            arr[index++] = i;
}

void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++)
        cout << arr[i] << " ";
    cout << endl;
}

int main() {
    int arr[] = {4, 2, 2, 8, 3, 3, 1};
    int n = sizeof(arr) / sizeof(arr[0]);

    countingSort(arr, n);

    cout << "Sorted array: ";
    printArray(arr, n);

    return 0;
}

File: kotlin/counting_sort.kt
fun countingSort(arr: IntArray): IntArray {
    if (arr.isEmpty()) return arr

    // Step 1: Find the maximum element to determine range
    val max = arr.maxOrNull() ?: return arr

    // Step 2: Create a count array
    val count = IntArray(max + 1)

    // Step 3: Count the occurrences of each number
    for (num in arr) {
        count[num]++
    }

    // Step 4: Reconstruct the sorted array
    var index = 0
    for (i in count.indices) {
        repeat(count[i]) {
            arr[index++] = i
        }
    }

    return arr
}

fun main() {
    print("Enter number of elements: ")
    val n = readln().toInt()

    val arr = IntArray(n)
    println("Enter elements:")
    for (i in 0 until n) {
        arr[i] = readln().toInt()
    }

    println("Original array: ${arr.joinToString(" ")}")
    val sorted = countingSort(arr)
    println("Sorted array:   ${sorted.joinToString(" ")}")
}
